File: solutions/typescript/bob/2/bob.ts
export default class Bob {
  hey(text: string): string {
    text = text.trim();
    const isYelling = text === text.toUpperCase() && /[A-Z]+/.test(text);
    const isAsking = text.endsWith("?");
    const isSayingNothing = text === "";

    if (isYelling && isAsking) return "Calm down, I know what I'm doing!";
    if (isSayingNothing) return "Fine. Be that way!";
    if (isYelling) return "Whoa, chill out!";
    if (isAsking) return "Sure.";
    return "Whatever.";
  }
}

File: solutions/java/bob/2/src/main/java/Bob.java
import java.util.regex.Pattern;
import java.util.function.Predicate;

class Bob {

    private final static Pattern isAlpha = Pattern.compile("[a-zA-Z]");
    private final static Predicate<String> isShout = msg -> isAlpha.matcher(msg).find() && msg == msg.toUpperCase();

    String hey(String input) {
        input = input.trim();
        if (input.isEmpty()) {
            return "Fine. Be that way!";            
        }
        
        var isShouting = isShout.test(input);
        var isQuestioning = input.endsWith("?");
        if (isQuestioning) {
            if (isShouting) {
                return "Calm down, I know what I'm doing!";
            }
            return "Sure.";
        }
        if (isShouting) {
            return "Whoa, chill out!";
        }
        return "Whatever.";
    }
}